aten/src/ATen/native/cuda/ScatterGatherKernel.cu

```diff
@@ -0,0 +1,167 @@
+#include <ATen/native/TensorAdvancedIndexing.h>
+
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+
+#include <ATen/native/ScatterGatherShapeChecks.h>
+#include <ATen/native/ReduceOpsUtils.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cuda/Loops.cuh>
+#include <ATen/core/Array.h>
+
+namespace at { namespace native {
+
+// The kernels are implemented on an opaque,
+// self-aligned type of the correct size,
+// to avoid redundant kernels for different types
+// of the same size.
+template <int N> struct alignas(N) OpaqueType { char data[N]; };
+
+template <int N>
+static OffsetCalculator<N> sg_make_offset_calculator(const TensorIterator& iter) {
+  AT_ASSERT(N <= iter.ntensors());
+  std::array<const int64_t*, N> strides;
+  for (int i = 0; i < N; ++i) {
+    strides[i] = iter.strides(i).data();
+  }
+  return OffsetCalculator<N>(iter.ndim(), iter.shape().data(), strides.data());
+}
+
+template <bool is_scatter_like, typename scalar_t>
+struct _cuda_scatter_gather_internal_kernel {
+  template <typename func_t>
+  void operator() (
+    TensorIterator& iter,
+    int64_t index_size,
+    int64_t index_stride,
+    const func_t& f
+  ) {
+    if (iter.numel() == 0) {
+      return;
+    }
+
+    if (!iter.can_use_32bit_indexing()) {
+      for (auto& sub_iter : iter.with_32bit_indexing()) {
+        _cuda_scatter_gather_internal_kernel<is_scatter_like, scalar_t>()(
+          sub_iter, index_size, index_stride, f
+        );
+      }
+      return;
+    }
+
+    char* self_ptr = (char*)iter.data_ptr(0);
+    char* src_ptr = (char*)iter.data_ptr(1);
+    char* index_ptr = (char*)iter.data_ptr(2);
+
+    auto offset_calc = sg_make_offset_calculator<3>(iter);
+    legacy::launch_kernel<launch_size_nd, launch_bound2>(iter.numel(), [=]__device__(int i) {
```

Comment:
legacy::launch_kernel is going to be removed soon, please see discussion in #35429

Reply:
Well, I guess it will require me writing my own kernel? The same kernel will be needed for index_put as well...

Reply:
Will `modern::launch_unrolled_kernel` will be sufficient here?

Reply:
None of the `gpu_` kernels seem to be suitable here. Shall I just write my own kernel? Looks like `launch_kernel` is being modified here in [34749](https://github.com/pytorch/pytorch/pull/34749). Shall I wait until it is landed?

Reply:
Yes, you can write your own kernel. #34749 may take a while to land because it is failing for complex, and the fix is pretty involved.